src/storage-controller/src/healthcheck.rs

```diff
@@ -122,97 +282,57 @@ where
+    /// `IntrospectionStatusUpdate`.
     ///
-    /// Panics if the sink status history collection is not registered in `introspection_ids`
-    pub async fn append_sink_updates(&mut self, updates: Vec<RawStatusUpdate>) {
-        self.append_updates(updates, IntrospectionType::SinkStatusHistory)
-            .await;
-    }
-
-    async fn drop_items(
-        &mut self,
-        ids: Vec<GlobalId>,
-        ts: DateTime<Utc>,
-        type_: IntrospectionType,
-    ) {
-        self.append_updates(
-            ids.iter()
-                .map(|id| RawStatusUpdate {
-                    id: *id,
-                    ts,
-                    status_name: "dropped".to_string(),
-                    error: None,
-                    hints: Default::default(),
-                    namespaced_errors: Default::default(),
-                })
-                .collect(),
-            type_,
-        )
-        .await;
-        // Note that don't remove the statuses from `previous_statuses`, as they will never be
-        // cleared from the history table
-        //
-        // TODO(guswynn): clean up dropped sources/sinks from status history tables, at some point.
-    }
-    /// Marks the sources as dropped.
-    ///
-    /// # Panics
-    ///
-    /// Panics if the source status history collection is not registered in `introspection_ids`
-    pub async fn drop_sources(&mut self, ids: Vec<GlobalId>, ts: DateTime<Utc>) {
-        self.drop_items(ids, ts, IntrospectionType::SourceStatusHistory)
-            .await;
-    }
+    /// Also tracks the value of the last-produced value for the inferred
+    /// `GlobalId` if appropriate.
+    fn should_produce_row(&mut self, type_: IntrospectionType, new: &Row) -> bool {
+        let type_: &IntrospectionStatusUpdate = match type_.try_into() {
+            Ok(type_) => type_,
+            // Treat all IntrospectionTypes without an IntrospectionStatusUpdate
+            // impl as append-only
+            Err(()) => return true,
+        };
+
+        let key = type_.global_id_idx;
+        let this_id = new
+            .iter()
+            .nth(key)
+            .expect("schema has not changed")
+            .unwrap_str();
+        let this_id = GlobalId::from_str(this_id).expect("GlobalIds must be parseable from datums");
+        let new_val = new.iter().nth(type_.value_idx).expect("schema unchanged");
+
+        let produce = match self.previous_statuses.get_mut(&this_id) {
+            None => {
+                let row = Row::pack_slice(&[new_val]);
+                self.previous_statuses.insert(this_id, row);
+                true
+            }
+            Some(prev) => {
+                let prev_val = prev.iter().next().expect("prev value always one datum");
+                let produce = (type_.produce_new_value)(prev_val, new_val);
+                if produce {
+                    let mut row = prev.packer();
+                    row.push(new_val);
```

Comment:
for a second I thought we had to put this back into `self.previous_statuses`. But this is mutating in-place, cool!